### PR TITLE
8290432: C2 compilation fails with assert(node->_last_del == _last) failed: must have deleted the edge just produced

### PR DESCRIPTION
**Pull request by y1yang0**
Hi, can I have a review for this patch? [JDK-8273585](https://bugs.openjdk.org/browse/JDK-8273585) recognized the form of `Phi->CastII->AddI` as additional parallel induction variables. In the following program:
```java
class Test {
    static int dontInline() {
        return 0;
    }

    static long test(int val, boolean b) {
        long ret = 0;
        long dArr[] = new long[100];
        for (int i = 15; 293 > i; ++i) {
            ret = val;
            int j = 1;
            while (++j < 6) {
                int k = (val--);
                for (long l = i; 1 > l; ) {
                    if (k != 0) {
                        ret += dontInline();
                    }
                }
                if (b) {
                    break;
                }
            }
        }
        return ret;
    }

    public static void main(String[] args) {
        for (int i = 0; i < 1000; i++) {
            test(0, false);
        }
    }
}
```
`val` is incorrectly matched with the new parallel IV form:
image
And C2 further replaces it with newly added nodes, which finally leads the crash:
image

I think we can add more constraints to the new form. The form of `Phi->CastXX->AddX` appears when using Preconditions.checkIndex, and it would be recognized as additional IV when 1) Phi != phi2, 2) CastXX is controlled by RangeCheck(to reflect changes in Preconditions checkindex intrinsic)

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed (2 reviews required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer), 1 [Author](https://openjdk.org/bylaws#author))

### Issue
 * [JDK-8290432](https://bugs.openjdk.org/browse/JDK-8290432): C2 compilation fails with assert(node->_last_del == _last) failed: must have deleted the edge just produced

### Reviewers
 * [Vladimir Kozlov](https://openjdk.org/census#kvn) (@vnkozlov - **Reviewer**)
 * [Tobias Hartmann](https://openjdk.org/census#thartmann) (@TobiHartmann - **Reviewer**)
 * [Christian Hagedorn](https://openjdk.org/census#chagedorn) (@chhagedorn - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk pull/9695/head:pull/9695` \
`$ git checkout pull/9695`

Update a local copy of the PR: \
`$ git checkout pull/9695` \
`$ git pull https://git.openjdk.org/jdk pull/9695/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 9695`

View PR using the GUI difftool: \
`$ git pr show -t 9695`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/9695.diff">https://git.openjdk.org/jdk/pull/9695.diff</a>

</details>

### CODE REVIEW

File: src/hotspot/share/opto/loopnode.cpp
Comment (vnkozlov) [inline src/hotspot/share/opto/loopnode.cpp:3682]: You can use `&&` instead of nested condition.

File: src/hotspot/share/opto/loopnode.cpp
Comment (vnkozlov) [inline src/hotspot/share/opto/loopnode.cpp:3679]: Since you touch near code, please update style in these lines too to use `{}`.

**Review by vnkozlov — COMMENTED** [review]
From your test I assume internal loop `for (long l=` is processed. But `val` is updated outside it. Its `AddI` node should not be part of loop's body. Could you explain?

**Comment by kelthuzadx** [comment]
Hi @vnkozlov, thanks for your review. Here is the (almost) complete IR before replacing additional parallel IVs:
image

Before that, some loop optimizations are applied: `[15,293)` and `[2,6)` are recognized as counted loops, and the latter loop is maximally unrolled. Loop unswitching is applied when C2 saw `if (b)`. The program looks like:
```java
static long test(int val, boolean b) {
    long ret = 0;
    long dArr[] = new long[100];
    if (b) {
        for (int i = 15; 293 > i; ++i) {
            ret = val;
            val--;
        }
    } else {
        for (int i = 15; 293 > i; ++i) {
            ret = val;
            val-=4
        }
    }
    return ret;
}
```
Their corresponding IR are as follows:
image

`Phi#564` and `Phi#125` represent `ret` in two unswitched loops, respectively. `val` is actually part of loop body.

File: src/hotspot/share/opto/loopnode.cpp
Comment (kelthuzadx) [reply src/hotspot/share/opto/loopnode.cpp]: The whole function is formatted.

File: src/hotspot/share/opto/loopnode.cpp
Comment (vnkozlov) [inline src/hotspot/share/opto/loopnode.cpp:3689]: Thank you for updating comment. But it still do not say which issue you are trying to avoid. Yes, this pattern can be generated by inline_preconditions_checkIndex.
Currently your check skip search for parallel IV if CastII is NOT controlled by RangeCheck!
But the comment in the BODY of check talks about using parallel IV.
It is contradicting.
Aslo `1) Phi != phi2` says nothing to me - your check does not have such condition.

Comment (vnkozlov) [reply src/hotspot/share/opto/loopnode.cpp:3689]: Why not just say that we should skip `AddI->CastII->Phi` case if CastII is not controlled by local RangeCheck.

**Comment by vnkozlov** [comment]
About test's loops. I see that C2 recognized that `for (long l = i; 1 > l; )` is not executed because `i >= 15`.
Then `while (++j < 6)` will be executed only once if `b == true` - C2 unswitch it with result of method code you showed.
Good.

BUT. `val` is `long` type why we are trying to use it as IV instead of existing `int i`? I think the issue is not in Cast node. There is mixture of `long` and `int` types here which confuses loop optimizer. What if `val` was `int` type?

File: src/hotspot/share/opto/loopnode.cpp
Comment (kelthuzadx) [reply src/hotspot/share/opto/loopnode.cpp:3689]: > Why not just say that we should skip `AddI->CastII->Phi` case if CastII is not controlled by local RangeCheck.

Yes, this is exactly what I want to say. Maybe I should update comments.

**Comment by kelthuzadx** [comment]
> BUT. `val` is `long` type. Why we are trying to use it as IV instead of existing `int i`? I think the issue is not in Cast node. There is mixture of `long` and `int` types here which confuses loop optimizer. What if `val` was `int` type?

val is type of int!

File: src/hotspot/share/opto/loopnode.cpp
Comment (TobiHartmann) [inline src/hotspot/share/opto/loopnode.cpp:3688]: In the valid case, isn't the ConstraintCast control input `incr2->in(1)->in(0)` the IfTrue projection of the RangeCheck?

I would remove the second line because it's not clear which "changes" in `LibraryCallKit::inline_preconditions_checkIndex` it is referring to.

```suggestion
```

**Review by TobiHartmann — COMMENTED** [review]
Could you explain why it's an issue in this case if the control input of the `CastII` is **not** a range check?

**Comment by TobiHartmann** [comment]
I think we should add an IR verification test for [JDK-8273585](https://bugs.openjdk.org/browse/JDK-8273585).

**Comment by vnkozlov** [comment]
> > BUT. `val` is `long` type. Why we are trying to use it as IV instead of existing `int i`? I think the issue is not in Cast node. There is mixture of `long` and `int` types here which confuses loop optimizer. What if `val` was `int` type?
> 
> val is type of int! And we are not trying to use it instead of existing `int i`, we are using both of them as IV.



> > BUT. `val` is `long` type. Why we are trying to use it as IV instead of existing `int i`? I think the issue is not in Cast node. There is mixture of `long` and `int` types here which confuses loop optimizer. What if `val` was `int` type?
> 
> val is type of int! And we are not trying to use it instead of existing `int i`, we are using both of them as IV.

Right. I mixed `ret` long type with `val`.

**Comment by vnkozlov** [comment]
I still think the issue is in some other place. Your change just avoiding the case which triggers it. The java code is legal and compiling next code did not trigger nay issues:
```
    static long test(int val, boolean b) {
         long ret = 0;
         for (int i = 15; 293 > i; ++i) {
             ret = val;
             val--;
         }
         return ret;
     }
```
Actually I was not able to reproduce the issue with included test.

**Comment by kelthuzadx** [comment]
> I still think the issue is in some other place.

Why? Can you clarify more? The root cause seems clear: the unexpected `Add->CastII->Phi` pattern was matched due to changes in [JDK-8273585](https://bugs.openjdk.org/browse/JDK-8273585) and finally leads the crash.

**Comment by kelthuzadx** [comment]
> I think we should add an IR verification test for [JDK-8273585](https://bugs.openjdk.org/browse/JDK-8273585).

Yes, we need a verification test for it. I'll do this later.

**Comment by TobiHartmann** [comment]
> Why? Can you clarify more?

As I mentioned above, I don't understand how your newly added condition is supposed to work.

**Comment by TobiHartmann** [comment]
Comment to keep this open. @kelthuzadx, any update on this?

**Comment by kelthuzadx** [comment]
> > Why? Can you clarify more?
> 
> As I mentioned above, I don't understand how your newly added condition is supposed to work.

@vnkozlov @TobiHartmann  I reviewed the above comments again. I think the proposed fix is good. According to my analysis at https://github.com/openjdk/jdk/pull/9695#issuecomment-1206221152 , we have a parallel IV which formed as `Add->CastII->Phi`, the birth of the form looks good, and this pattern is only legal after [JDK-8273585](https://bugs.openjdk.org/browse/JDK-8273585). In that case, `Add->CastII->Phi` parallel IV was generated by Preconditions.checkIndex, CastII was controlled by  ConstraintCast. In the proposed fix, I'm trying to avoid to recognize similar pattern if CastII was not controlled by ConstraintCast.



I reviewed the previous discussion and I think this fix is ok. According to my analysis(https://github.com/openjdk/jdk/pull/9695#issuecomment-1206221152 ), the entire IR(Add-CastII-Phi) generation is reasonable, and this format is only valid after [JDK-8273585](https://bugs.openjdk.org/browse/JDK-8273585). In that case, CastII is controlled by ConstraintCast, so in the current patch, we refuse to recognize other cases besides this. As far as this fix itself is concerned, I think the whole process is reasonable. The only thing I'm not sure about is whether this Add-CastII-Phi form is really reasonable as an IV, which is why I think adding an IR verification test for [JDK-8273585](https://bugs.openjdk.org/browse/JDK-8273585) is reasonable, I will go back to investigate that again after this patch, what do you think?

File: src/hotspot/share/opto/loopnode.cpp
Comment (TobiHartmann) [reply src/hotspot/share/opto/loopnode.cpp:3688]: It's still not clear to me how the `incr2->in(1)->in(0)->is_RangeCheck()` condition can ever be true. How can the control input of a ConstraintCast be a RangeCheck? Shouldn't there be a projection node in-between?

File: test/hotspot/jtreg/compiler/c2/TestUnexpectedParallelIV.java
Comment (TobiHartmann) [inline test/hotspot/jtreg/compiler/c2/TestUnexpectedParallelIV.java:28]: This test does not reproduce the issue for me, whereas [Test-2.java](https://bugs.openjdk.org/secure/attachment/100710/Test-2.java) still works.

File: src/hotspot/share/opto/loopnode.cpp
Comment (kelthuzadx) [reply src/hotspot/share/opto/loopnode.cpp:3688]: I missed this comment before... I need incr2->in(1)->in(0)->in(0)->is_RangeCheck()

File: test/hotspot/jtreg/compiler/c2/TestUnexpectedParallelIV.java
Comment (kelthuzadx) [reply test/hotspot/jtreg/compiler/c2/TestUnexpectedParallelIV.java:28]: You can reset to JDK-8273585, TestUnexpectedParallelIV.java reproduces the crash. Test-2.java can always reproduce without resetting commits. I added both of them as test cases.

**Review by TobiHartmann — COMMENTED** [review]
The new version looks reasonable to me but please merge the two tests into one. There is also a jcheck whitespace error.

If you don't intend to add an IR verification test for [JDK-8273585](https://bugs.openjdk.org/browse/JDK-8273585) with this fix, please file at least a follow-up enhancement.

**Comment by kelthuzadx** [comment]
I filed adding IR verification test as https://bugs.openjdk.org/browse/JDK-8297307.

**Review by TobiHartmann — APPROVED** [review]
Looks good to me.

**Review by TobiHartmann — APPROVED** [review]
Looks good to me.

**Comment by kelthuzadx** [comment]
@vnkozlov Can I have a second review? Thanks!

**Review by vnkozlov — COMMENTED** [review]
I may ask to do performance testing for this change too before approval.

**Comment by y1yang0** [comment]
> I may ask to do our internal performance testing for this change too before approval.

Do you mean... oracle internal performance testing or opensource microbenchmark inside openjdk? For the former one, may I ask your help to do that? Thanks. (For the later one, I tested JDK-8273585 attached test, and it works well. After this patch, it takes 8s.)

**Comment by TobiHartmann** [comment]
I'll run some performance testing in our system and report back once it passed.

**Comment by TobiHartmann** [comment]
Performance results look good (no measurable difference). A second review would be good though.

**Comment by chhagedorn** [comment]
Do we really need this bailout? From your very [first picture](https://user-images.githubusercontent.com/5010047/182059398-fc5204bc-8d95-4e3e-8c66-15776af457b8.png), it does not seem wrong to replace `125 Phi` - the only problem is that `465 CastII` also happens to have `432 CountedLoop` as control input. This leads to the crash because we are removing two outputs of `432 CountedLoop` at once which is not valid and previously unexpected (before [JDK-8273585](https://bugs.openjdk.org/browse/JDK-8273585)) . Couldn't we instead just update `replace_parallel_iv()` in such a way that it can handle this particular case of removing two outputs at once from the counted loop?

**Comment by y1yang0** [comment]
> it does not seem wrong to replace 125 Phi - the only problem is that 465 CastII also happens to have 432 CountedLoop as control input. 

If replacing Phi#125 is acceptable, we need to reach a consensus that the form of `Add->CastII->Phi` can be considered as parallel IV, which is not valid before [JDK-8273585](https://bugs.openjdk.org/browse/JDK-8273585). Prior to that, we only considered variables of the form `X += constant`(`Add->Phi`) to be IV. 

Do we have such a consensus that the form of `Add->CastII->Phi` is considered as parallel IV? A conservative approach would be a compilation bailout, as this patch does.

**Review by chhagedorn — APPROVED** [review]
> If replacing Phi#125 is acceptable, we need to reach a consensus that the form of Add->CastII->Phi can be considered as parallel IV

I think it should be safe and we've already done that since JDK-8273585 and haven't seen a crash related to that idea. But given how close we are to the fork, I suggest to go with your bailout fix for JDK 20 which is safer (and performance testing done by Tobias looks good). In this way, we really only optimize the pattern originally intended in JDK-8273585.

For the general case of allowing any cast node, I suggest to file an RFE and investigate again if that is possible/correct. I have the feeling that it is but it would be better to defer that to JDK 21. What do you think?

Thanks,
Christian

**Comment by TobiHartmann** [comment]
That makes sense, you can re-use/extend [JDK-8297307](https://bugs.openjdk.org/browse/JDK-8297307) for that.

Vladimir should also have a look at this again.

**Comment by y1yang0** [comment]
> For the general case of allowing any cast node, I suggest to file an RFE and investigate again if that is possible/correct.
> 
> Thanks, Christian

I think it’s really reasonable. As Tobias said, we can reuse https://bugs.openjdk.org/browse/JDK-8297307 for this purpose. Thanks!

**Review by vnkozlov — APPROVED** [review]

**Pull request closed without merging**
